hoodie-client/src/main/java/com/uber/hoodie/table/HoodieMergeOnReadTable.java

```diff
@@ -388,6 +397,116 @@ public HoodieCompactionPlan scheduleCompaction(JavaSparkContext jsc, String inst
+        Map<String, HoodieRollingStat> partitionRollingStats =
+            rollingStatMetadata.getPartitionToRollingStats().get(partitionPath);
+        if (partitionRollingStats != null) {
+          HoodieRollingStat rollingStatForFile = partitionRollingStats.get(fileSlice.getFileId());
+          if (rollingStatForFile != null) {
+            long inserts = rollingStatForFile.getInserts();
+            long totalSize = averageRecordSize * inserts;
+            return totalSize;
+          }
+        }
+      }
+      // In case Rolling Stats is not present, fall back to sizing log files based on heuristics
+      if (!fileSlice.getDataFile().isPresent()) {
+        return convertLogFilesSizeToExpectedParquetSize(fileSlice.getLogFiles().collect(Collectors.toList()));
+      } else {
+        return fileSlice.getDataFile().get().getFileSize() + convertLogFilesSizeToExpectedParquetSize(fileSlice
+            .getLogFiles().collect(Collectors.toList()));
+      }
+    }
+
+    private boolean isSmallFile(String partitionPath, FileSlice fileSlice) {
+      long totalSize = getTotalFileSize(partitionPath, fileSlice);
+      if (totalSize < config.getParquetMaxFileSize()) {
+        return true;
+      }
+      return false;
+    }
+
+    // TODO (NA) : Make this static part of utility
+    @VisibleForTesting
+    public long convertLogFilesSizeToExpectedParquetSize(List<HoodieLogFile> hoodieLogFiles) {
+      long totalSizeOfLogFiles = hoodieLogFiles.stream().map(hoodieLogFile -> hoodieLogFile.getFileSize().get())
+          .reduce((a, b) -> (a + b)).orElse(0L);
+      // Here we assume that if there is no base parquet file, all log files contain only inserts.
+      // We can then just get the parquet equivalent size of these log files, compare that with
+      // {@link config.getParquetMaxFileSize()} and decide if there is scope to insert more rows
+      long logFilesEquivalentParquetFileSize = (long) (totalSizeOfLogFiles * config
+          .getLogFileToParquetCompressionRatio());
+      return logFilesEquivalentParquetFileSize;
+    }
+  }
+
+  private Map<HeaderMetadataType, String> generateHeader(String commit) {
+    // generate metadata
+    Map<HeaderMetadataType, String> header = Maps.newHashMap();
+    header.put(HeaderMetadataType.INSTANT_TIME, metaClient.getActiveTimeline().lastInstant().get().getTimestamp());
+    header.put(HeaderMetadataType.TARGET_INSTANT_TIME, commit);
+    header.put(HeaderMetadataType.COMMAND_BLOCK_TYPE, String.valueOf(HoodieCommandBlockTypeEnum.ROLLBACK_PREVIOUS_BLOCK
+        .ordinal()));
+    return header;
+  }
+
+  private HoodieRollbackStat rollback(HoodieIndex hoodieIndex, String partitionPath, String commit,
+      HoodieCommitMetadata commitMetadata, final Map<FileStatus, Boolean> filesToDeletedStatus,
+      Map<FileStatus, Long> filesToNumBlocksRollback, Set<String> deletedFiles) {
+    // The following needs to be done since GlobalIndex at the moment does not store the latest commit time.
```

Comment:
@n3nash can you compile a list of these one-offs into a ticket.. we need to fish them all out eventually.